File: src/main/java/com/github/ofrostdev/api/utils/SectionBuilder.java
package com.github.ofrostdev.api.utils;

import com.github.ofrostdev.api.utils.items.ItemBuilder;
import org.bukkit.*;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.EntityType;
import org.bukkit.inventory.ItemStack;

import java.lang.reflect.Constructor;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

// fork by Frost
public class SectionBuilder<T> {

    private static final Map<Class<?>, Adapter<?>> CLASS_ADAPTERS = new HashMap<>();

    static {
        CLASS_ADAPTERS.put(ItemStack.class, new ItemAdapter());
        CLASS_ADAPTERS.put(String.class, new StringAdapter());
        CLASS_ADAPTERS.put(Material.class, new MaterialAdapter());
        CLASS_ADAPTERS.put(Location.class, new LocationAdapter());
        CLASS_ADAPTERS.put(StringList.class, new ListAdapter<>(new StringAdapter()));
        CLASS_ADAPTERS.put(ItemList.class, new ListAdapter<>(new ItemAdapter()));
        CLASS_ADAPTERS.put(Sound.class, new EnumAdapter<>(Sound.class));
        CLASS_ADAPTERS.put(EntityType.class, new EnumAdapter<>(EntityType.class));
    }

    private final ConfigurationSection mainSection;
    private final Class<T> clazz;
    private final Constructor<T> constructor;
    private final Map<String, Class<?>> parameters;
    private final Map<String, Adapter<?>> parametersAdapters;

    public static <E> SectionBuilder<E> of(Class<E> clazz, ConfigurationSection section) {
        return new SectionBuilder<E>(clazz, section);
    }

    private SectionBuilder(Class<T> clazz, ConfigurationSection mainSection) {
        this.mainSection = mainSection;
        this.clazz = clazz;
        this.constructor = (Constructor<T>) clazz.getConstructors()[0];
        this.parameters = new LinkedHashMap<>();
        this.parametersAdapters = new HashMap<>();
    }


    public SectionBuilder<T> parameter(String key, Class<?> valueClass) {
        this.parameters.put(key, valueClass);
        return this;
    }

    public SectionBuilder<T> parameter(String key, Class<?> valueClass, Adapter<?> adapter) {
        this.parameters.put(key, valueClass);
        this.parametersAdapters.put(key, adapter);
        return this;
    }

    public SectionBuilder<T> adapter(Class<?> clazz, Adapter<?> adapter) {
        CLASS_ADAPTERS.put(clazz, adapter);
        return this;
    }


    public List<T> build() {
        if (constructor.getParameterCount() - 1 != parameters.size()) { // primeiro argumento é sempre a key, e não precisa ser passado pelo método parameter(String, Class)
            throw new IllegalArgumentException("Constructor has " + (constructor.getParameterCount() - 1) + " parameters, but it was passed " + parameters.size() + " parameters.");
        }
        List<T> toReturn = new ArrayList<>();
        for (String key : mainSection.getKeys(false)) {
            try {
                ConfigurationSection section = mainSection.getConfigurationSection(key);
                List<Object> constructorParameters = new ArrayList<>();
                constructorParameters.add(key);
                for (Map.Entry<String, Class<?>> entry : parameters.entrySet()) {
                    String parameter = entry.getKey();
                    Class<?> parameterClass = entry.getValue();
                    Object object = section.get(parameter);
                    Adapter<?> adapter = parametersAdapters.getOrDefault(parameter, CLASS_ADAPTERS.get(parameterClass));
                    constructorParameters.add(adapter == null ? object : adapter.supply(object));
                }
                T instance = constructor.newInstance(constructorParameters.toArray());
                toReturn.add(instance);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return toReturn;
    }

    public <E> Map<E, T> build(Function<T, E> extractor) {
        return build().stream().collect(Collectors.toMap(extractor, Function.identity()));
    }

    public static interface Adapter<A> {
        A supply(Object object);
    }

    public static class ItemAdapter implements Adapter<ItemStack> {

        private final static MaterialAdapter MATERIAL_ADAPTER = new MaterialAdapter();
        private final static StringAdapter STRING_ADAPTER = new StringAdapter();

        /* O único atributo necessário é o material
            material: STONE
            #material: http://textures.minecraft.net/texture/fa8887814578ce7c540d7ab8cc6b2a2e22a7492cc86c65a7e839c887b2ed62 # caso tenha essa propriedade, o item automaticamente ira ser uma skull com a textura setada (pode ser um nick também)
            display-name: "&bFrost"
            amount: 1
            lore:
             - "&ePlugins"
            enchants:
              - DAMAGE_ALL:10
            nbts: # %uuid_random%
              - "key:value"
            glow: false
            dont-stack: true
         */
        @Override
        public ItemStack supply(Object object) {
            ConfigurationSection section = (ConfigurationSection) object;

            ItemBuilder ib;
            String materialStr = section.getString("material");
            int amount = section.isSet("amount") ? section.getInt("amount") : 1;
            String displayName = section.isSet("display-name") ? STRING_ADAPTER.supply(section.getString("display-name")) : null;
            List<String> lore = section.isSet("lore") ? section.getStringList("lore").stream().map(STRING_ADAPTER::supply).collect(Collectors.toList()) : null;
            boolean glow = section.isSet("glow") && section.getBoolean("glow");
            boolean dontstack = section.isSet("dont-stack") && section.getBoolean("dont-stack");
            Map<Enchantment, Integer> enchants = null;

            if (section.isSet("enchants")) {
                enchants = section.getStringList("enchants").stream()
                        .map(s -> s.split(":"))
                        .collect(Collectors.toMap(a -> Enchantment.getByName(a[0]), a -> Integer.parseInt(a[1].trim())));
            }

            if (materialStr != null) {
                if (materialStr.startsWith("http://textures.minecraft.net/texture/") ||
                        materialStr.startsWith("https://textures.minecraft.net/texture/")) {
                    ib = new ItemBuilder(materialStr); // skull com textura
                } else {
                    Material material;
                    short data = 0;
                    if (materialStr.contains(":")) {
                        String[] split = materialStr.split(":");
                        try {
                            material = Material.valueOf(split[0].toUpperCase());
                            data = Short.parseShort(split[1]);
                        } catch (Exception e) {
                            material = Material.STONE;
                            data = 0;
                        }
                    } else {
                        try {
                            material = Material.valueOf(materialStr.toUpperCase());
                        } catch (Exception e) {
                            material = Material.STONE;
                        }
                    }
                    ib = new ItemBuilder(material, amount, data);
                }
            } else {
                ib = new ItemBuilder(Material.STONE, amount);
            }

            if (displayName != null) ib.setName(displayName);
            if (lore != null) ib.setLore(lore);
            if (glow) ib.setGlowing(true);
            if (dontstack) ib.setNBTString("frostkkj", UUID.randomUUID().toString());

            if (enchants != null)
                enchants.forEach(ib::addEnchant);

            if (section.isSet("nbts")) {
                List<String> nbtList = section.getStringList("nbts");
                for (String nbtEntry : nbtList) {
                    String[] split = nbtEntry.split(":", 2);
                    if (split.length != 2) continue;
                    ib.setNBTString(split[0].trim(), split[1].trim().replace("%uuid_random%", UUID.randomUUID().toString()));
                }
            }

            ib.setAmount(amount);
            return ib.build();
        }
    }

    private static class StringAdapter implements Adapter<String> {

        @Override
        public String supply(Object object) {
            return ChatColor.translateAlternateColorCodes('&', (String) object);
        }
    }

    private static class MaterialAdapter implements Adapter<Material> {

        @Override
        public Material supply(Object object) {
            String value = object.toString();
            return isNumber(value) ? Material.getMaterial(Integer.parseInt(value)) : Material.valueOf(value.toUpperCase());
        }

        private boolean isNumber(String string) {
            try {
                int i = Integer.parseInt(string);
                return true;
            } catch (NumberFormatException e) {
                return false;
            }
        }

    }

    public static class ListAdapter<A> implements Adapter<List<A>> {

        private final Adapter<A> adapter;

        public ListAdapter(Adapter<A> adapter) {
            this.adapter = adapter;
        }

        @Override
        public List<A> supply(Object object) {
            if (object instanceof ConfigurationSection) {
                ConfigurationSection section = (ConfigurationSection) object;
                return section.getKeys(false).stream().map(section::getConfigurationSection).map(adapter::supply).collect(Collectors.toList());
            } else {
                List<Object> list = (List<Object>) object;
                return list.stream().map(adapter::supply).collect(Collectors.toList());
            }
        }
    }

    public static class LocationAdapter implements Adapter<Location> {

        @Override
        public Location supply(Object object) {
            String value = (String) object;
            if (value == null || !value.contains(":")) return null;

            String[] parts = value.split(":");
            if (parts.length < 6) return null;

            World w = Bukkit.getServer().getWorld(parts[0]);
            if (w == null) return null;

            double x = Double.parseDouble(parts[1]);
            double y = Double.parseDouble(parts[2]);
            double z = Double.parseDouble(parts[3]);
            float yaw = Float.parseFloat(parts[4]);
            float pitch = Float.parseFloat(parts[5]);

            return new Location(w, x, y, z, yaw, pitch);
        }
    }

    public static class EnumAdapter<A extends Enum<A>> implements Adapter<A> {

        private final Class<A> enumClass;

        public EnumAdapter(Class<A> aEnum) {
            this.enumClass = aEnum;
        }

        @Override
        public A supply(Object object) {
            return Enum.valueOf(enumClass, (String) object);
        }
    }

    public static class ItemList {
    }

    public static class StringList {
    }
}

File: src/main/kotlin/com/github/ofrostdev/api/utils/common/fileconfiguration/SectionBuilder.kt
package com.github.ofrostdev.api.utils.common.fileconfiguration

import com.github.ofrostdev.api.utils.common.item.ItemBuilder
import org.bukkit.*
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.enchantments.Enchantment
import org.bukkit.entity.EntityType
import org.bukkit.inventory.ItemStack
import java.lang.reflect.Constructor
import java.util.*
import kotlin.collections.LinkedHashMap

@DslMarker
annotation class SectionBuilderDSL

@SectionBuilderDSL
class KSectionBuilder<T>(clazz: Class<T>, section: ConfigurationSection) {

    private val builder: SectionBuilder<T> = SectionBuilder.of(clazz, section)

    fun parameter(key: String, type: Class<*>, adapter: SectionBuilder.Adapter<*>? = null) {
        if (adapter != null) builder.parameter(key, type, adapter) else builder.parameter(key, type)
    }

    fun adapter(clazz: Class<*>, adapter: SectionBuilder.Adapter<*>) = builder.adapter(clazz, adapter)

    fun build(): List<T> = builder.build()

    fun <E> buildMap(extractor: (T) -> E): Map<E, T> = builder.buildMap(extractor)
}

fun <T> section(clazz: Class<T>, section: ConfigurationSection, block: KSectionBuilder<T>.() -> Unit): List<T> {
    val kBuilder = KSectionBuilder(clazz, section)
    kBuilder.block()
    return kBuilder.build()
}

class SectionBuilder<T> private constructor(
    private val mainSection: ConfigurationSection,
    private val clazz: Class<T>
) {

    private val constructor: Constructor<T> = clazz.constructors[0] as Constructor<T>
    private val parameters: LinkedHashMap<String, Class<*>> = LinkedHashMap()
    private val parametersAdapters: MutableMap<String, Adapter<*>> = mutableMapOf()

    companion object {
        private val CLASS_ADAPTERS: MutableMap<Class<*>, Adapter<*>> = mutableMapOf(
            ItemStack::class.java to ItemAdapter(),
            String::class.java to StringAdapter(),
            Material::class.java to MaterialAdapter(),
            Location::class.java to LocationAdapter(),
            StringList::class.java to ListAdapter(StringAdapter()),
            ItemList::class.java to ListAdapter(ItemAdapter()),
            Sound::class.java to EnumAdapter(Sound::class.java),
            EntityType::class.java to EnumAdapter(EntityType::class.java),
            ListEffect::class.java to EffectAdapter()
        )

        fun <E> of(clazz: Class<E>, section: ConfigurationSection) = SectionBuilder<E>(section, clazz)
    }

    fun parameter(key: String, type: Class<*>): SectionBuilder<T> {
        parameters[key] = type
        return this
    }

    fun parameter(key: String, type: Class<*>, adapter: Adapter<*>): SectionBuilder<T> {
        parameters[key] = type
        parametersAdapters[key] = adapter
        return this
    }

    fun adapter(clazz: Class<*>, adapter: Adapter<*>): SectionBuilder<T> {
        CLASS_ADAPTERS[clazz] = adapter
        return this
    }

    fun build(): List<T> {
        if (constructor.parameterCount - 1 != parameters.size) {
            throw IllegalArgumentException("Constructor has ${constructor.parameterCount - 1} parameters, but ${parameters.size} provided.")
        }

        return mainSection.getKeys(false).mapNotNull { key ->
            try {
                val section = mainSection.getConfigurationSection(key)
                val args = mutableListOf<Any?>(key)
                parameters.forEach { (param, type) ->
                    val value = section?.get(param)
                    val adapter = parametersAdapters[param] ?: CLASS_ADAPTERS[type]
                    args.add(adapter?.supply(value) ?: value)
                }
                constructor.newInstance(*args.toTypedArray())
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }
    }

    fun <E> buildMap(extractor: (T) -> E): Map<E, T> = build().associateBy(extractor)

    interface Adapter<A> { fun supply(obj: Any?): A }

    class ItemAdapter : Adapter<ItemStack> {
        private val stringAdapter = StringAdapter()
        override fun supply(obj: Any?): ItemStack {
            val section = obj as ConfigurationSection
            val materialStr = section.getString("material")
            val amount = section.getInt("amount", 1)
            val displayName = section.getString("display-name")?.let { stringAdapter.supply(it) }
            val lore = section.getStringList("lore")?.map(stringAdapter::supply)
            val glow = section.getBoolean("glow", false)
            val dontStack = section.getBoolean("dont-stack", false)
            val enchants = section.getStringList("enchantments").associate {
                val (enchant, level) = it.split(":")
                Enchantment.getByName(enchant) to level.toInt()
            }.takeIf { it.isNotEmpty() }

            val ib = if (materialStr != null) {
                if (materialStr.startsWith("http")) ItemBuilder(materialStr)
                else {
                    val material = try { Material.valueOf(materialStr.uppercase()) } catch (_: Exception) { Material.STONE }
                    ItemBuilder(material, amount)
                }
            } else ItemBuilder(Material.STONE, amount)

            displayName?.let { ib.setName(it) }
            lore?.let { ib.setLore(it) }
            if (glow) ib.setGlowing(true)
            if (dontStack) ib.setNBTString("frostkkj", UUID.randomUUID().toString())
            enchants?.forEach { (ench, lvl) -> ib.addEnchantment(ench, lvl) }
            section.getStringList("nbts").forEach {
                val (k, v) = it.split(":", limit = 2)
                ib.setNBTString(k, v.replace("%uuid_random%", UUID.randomUUID().toString()))
            }
            return ib.build()
        }
    }

    class StringAdapter : Adapter<String> { override fun supply(obj: Any?) = ChatColor.translateAlternateColorCodes('&', obj as String) }
    class MaterialAdapter : Adapter<Material> { override fun supply(obj: Any?) = if (obj.toString().toIntOrNull() != null) Material.getMaterial(obj.toString().toInt()) else Material.valueOf(obj.toString().uppercase()) }
    class LocationAdapter : Adapter<Location?> {
        override fun supply(obj: Any?): Location? {
            val value = obj as? String ?: return null
            val parts = value.split(":")
            if (parts.size < 6) return null
            val world = Bukkit.getWorld(parts[0]) ?: return null
            val x = parts[1].toDoubleOrNull() ?: return null
            val y = parts[2].toDoubleOrNull() ?: return null
            val z = parts[3].toDoubleOrNull() ?: return null
            val yaw = parts[4].toFloatOrNull() ?: return null
            val pitch = parts[5].toFloatOrNull() ?: return null
            return Location(world, x, y, z, yaw, pitch)
        }
    }

    class EnumAdapter<A : Enum<A>>(private val enumClass: Class<A>) : Adapter<A> {
        override fun supply(obj: Any?) = java.lang.Enum.valueOf(enumClass, obj as String)
    }

    class ListAdapter<A>(private val adapter: Adapter<A>) : Adapter<List<A>> {
        override fun supply(obj: Any?): List<A> {
            return when (obj) {
                is ConfigurationSection -> obj.getKeys(false).map { adapter.supply(obj.getConfigurationSection(it)) }
                is List<*> -> obj.map(adapter::supply)
                else -> emptyList()
            }
        }
    }

    class EffectAdapter : Adapter<List<EffectAdapter.EffectData>> {
        class EffectData(val name: String, val amplifier: Int, val duration: Int)
        override fun supply(obj: Any?): List<EffectData> {
            return (obj as? List<*>)?.mapNotNull {
                val parts = (it as? String)?.split(":") ?: return@mapNotNull null
                if (parts.size != 3) return@mapNotNull null
                try { EffectData(parts[0].uppercase(), parts[1].toInt(), parts[2].toInt()) } catch (_: Exception) { null }
            } ?: emptyList()
        }
    }

    class ItemList
    class ListEffect
    class StringList
}
